### PR TITLE
Add hww arbitration layer

### PR DESCRIPTION
**Pull request by conte91**
Add a layer of encapsulation to HWW packets, to handle data retransmission and arbitration (HWW_REQ_NEW, HWW_REQ_RETRY, HWW_REQ_CANCEL).

### CODE REVIEW

File: src/util.h [hunk 1]
```diff
@@ -127,4 +127,26 @@ bool safe_uint64_add(uint64_t* a, uint64_t b);
             fprintf(stdout, "%s:%d: " format "\n", FILENAME, __LINE__, __VA_ARGS__); \
     } while (0)
 
+/**
+ * Struct definining a rw buffer (buffer + length).
+ */
+typedef struct {
+    /** Data buffer. */
+    uint8_t* data;
```
Comment (benma) [inline src/util.h]: somewhere in here a `const` can be added to indicate that the data pointer can't be changed (but the contents can)

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py [hunk 2]
```diff
@@ -42,6 +42,27 @@
 
 HWW_CMD = 0x80 + 0x40 + 0x01
 
+
+class HwwRequestCode:
+    # New request.
+    HWW_REQ_NEW = b"\x00"
```
Comment (benma) [inline py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: prefixes not needed? unlike C those are luckily not in the global namespace :)

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py [hunk 3]
```diff
@@ -404,6 +425,66 @@ def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
...
+    # Send the message to the device, resend if the device is busy.
+    # Poll for a result until the query is completed.
+    def _execute_raw_query(self, msg: bytes, cid: int) -> bytes:
+        status = None
+        payload = None
+        msg_sent = False
+        while not msg_sent:
+            response = self._transport.query(HwwRequestCode.HWW_REQ_NEW + msg, HWW_CMD, cid)
+            assert len(response) != 0, "Unexpected response of length 0 from HWW stack."
+            status, payload = response[0:1], response[1:]
```
Comment (benma) [inline py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: `[:1]`

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py [hunk 4]
```diff
@@ -404,6 +425,66 @@ def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
...
+            if status != HwwResponseCode.HWW_RSP_BUSY:
+                msg_sent = True
+            else:
+                assert (
```
Comment (benma) [inline py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: a bit shorter/easier: `assert len(payload) == 0, ...`

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py [hunk 5]
```diff
@@ -404,6 +425,66 @@ def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
...
+            raise Exception("Unexpected NACK response from HWW stack.")
+
+        # The message has been sent. If we have a retry, poll the device until we're ready.
+        while status == HwwResponseCode.HWW_RSP_NOT_READY:
+            assert (
+                len(payload) == 0
+            ), "Unexpected payload of length {} with HWW_RSP_NOT_READY response.".format(
+                len(payload)
+            )
+            time.sleep(0.2)
```
Comment (benma) [inline py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: not imported, does this even work? :D

File: src/util.h (hunk 1 above)
Comment (conte91) [reply src/util.h]: Then one should just pass in a `const buffer_t`!

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py (hunk 5 above)
Comment (conte91) [reply py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: Got lost in the process! Fixed in 2756844.

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py (hunk 4 above)
Comment (conte91) [reply py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: Fixed 2756844

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py (hunk 3 above)
Comment (conte91) [reply py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: Fixed 2756844

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py (hunk 2 above)
Comment (conte91) [reply py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: Fixed 2756844

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py [hunk 6]
```diff
@@ -404,6 +425,66 @@ def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
...
+            if status not in [HwwResponseCode.HWW_RSP_NOT_READY, HwwResponseCode.HWW_RSP_ACK]:
+                # We should never receive a NACK unless some internal error occurs.
+                raise Exception(
+                    "Unexpected response from HWW stack during retry ({}).".format(status)
+                )
+        return payload
+
+    def _raw_query(self, msg: bytes) -> bytes:
+        """
+        Starting with v7.0.0, HWW messages are incapsulated
```
[the remark this answers is not in the document]
Comment (conte91) [reply py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: Removed Italianglish 2756844

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py [hunk 7]
```diff
@@ -404,6 +425,66 @@ def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
...
+            assert len(response) != 0, "Unexpected response of length 0 from HWW stack."
+            status, payload = response[0:1], response[1:]
+            if status != HwwResponseCode.HWW_RSP_BUSY:
+                msg_sent = True
+            else:
+                assert (
+                    len(payload) == 0
+                ), "Unexpected payload of length {} with HWW_RSP_BUSY response.".format(
+                    len(payload)
+                )
```
[the remark this answers is not in the document]
Comment (conte91) [reply py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: Fixed 2756844

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py [hunk 8]
```diff
@@ -404,6 +425,66 @@ def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
...
+        into an arbitration layer. The device can respond with
+        HWW_RSP_NOT_READY to indicate that we should poll it later.
+        """
+        cid = self._transport.generate_cid()
+        response = None
+        try:
+            response = self._execute_raw_query(msg, cid)
+        except KeyboardInterrupt:
+            # Send a cancel message.
+            response = self._transport.query(HwwRequestCode.HWW_REQ_RETRY, HWW_CMD, cid)
```
[the remark this answers is not in the document]
Comment (conte91) [reply py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: `HWW_CANCEL` had broken semantics - Fixed in 2756844 after discussion

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py [hunk 9]
```diff
@@ -404,6 +425,66 @@ def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
...
+    def _execute_raw_query(self, msg: bytes, cid: int) -> bytes:
+        status = None
+        payload = None
+        msg_sent = False
+        while not msg_sent:
+            response = self._transport.query(HwwRequestCode.HWW_REQ_NEW + msg, HWW_CMD, cid)
+            assert len(response) != 0, "Unexpected response of length 0 from HWW stack."
+            status, payload = response[0:1], response[1:]
+            if status != HwwResponseCode.HWW_RSP_BUSY:
+                msg_sent = True
```
[the remark this answers is not in the document]
Comment (conte91) [reply py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: Cleaned up a68e42b

File: src/util.h (hunk 1 above)
Comment (conte91) [reply src/util.h]: *Admits defeat* f82a689

File: py/bitbox02/bitbox02/communication/bitbox_api_protocol.py (hunk 2 above)
Comment (conte91) [reply py/bitbox02/bitbox02/communication/bitbox_api_protocol.py]: + 136a47f

File: src/hww.c [hunk 10]
```diff
@@ -163,11 +189,50 @@ static void _msg(const Packet* in_packet, Packet* out_packet, const size_t max_o
...
+    case HWW_REQ_RETRY:
+        /* We don't support async retries yet. */
+        response.status = HWW_RSP_NACK;
+        break;
+    default:
+        break;
+    }
+    out_packet->data_addr[0] = response.status;
+    if (response.status == HWW_RSP_ACK) {
+        out_packet->len = response.buffer.len + 1;
```
Comment (benma) [inline src/hww.c]: you can use this line regardless of the response status, right? response.buffer.len is 0 if nothing was written to it.

**Pull request merged**